### PR TITLE
More refactors for SQLAlchemy

### PR DESCRIPTION
**Pull request by svanscho**

### CODE REVIEW

File: src/gateway/metrics_collector.py [hunk 1]
```diff
@@ -427,17 +427,6 @@ def _run_system(self, metric_type):
                 except Exception as ex:
                     logger.error('Error loading disk eMMC metrics: {0}'.format(ex))
 
-                # get database metrics
```
Comment (svanscho) [inline src/gateway/metrics_collector.py:430]: assuming it's fine to leave these out

File: src/gateway/webservice.py [hunk 2]
```diff
@@ -2018,11 +2018,9 @@ def get_reset_dirty_flag(self):
         """
         energy_dirty = self._energy_dirty
         self._energy_dirty = False
-        orm_dirty = Database.get_dirty_flag()
```
Comment (svanscho) [inline src/gateway/webservice.py:2021]: should not be used for syncing any longer

File: src/gateway/mappers/schedule.py [hunk 3]
```diff
@@ -45,8 +45,8 @@ def orm_to_dto(orm_object):  # type: (Schedule) -> ScheduleDTO
                            arguments=arguments)
 
     @staticmethod
-    def dto_to_orm(schedule_dto):  # type: (ScheduleDTO) -> Schedule
-        schedule = Schedule.get_or_none(id=schedule_dto.id)
+    def dto_to_orm(db, schedule_dto):  # type: (Session, ScheduleDTO) -> Schedule
+        schedule = db.query(Schedule).where(Schedule.id == schedule_dto.id).one_or_none()
```
Comment (svanscho) [inline src/gateway/mappers/schedule.py:49]: move db to __init

File: src/gateway/models.py [hunk 4]
```diff
@@ -249,52 +249,53 @@ class Config(Base):
...
+    def get_entry(key, fallback):
+        # type: (str, T) -> T
+        """ Retrieves a setting from the DB, returns the argument 'fallback' when non existing """
+        key = key.lower()
+        if key in Config.CACHE:
+            data, expire_at = Config.CACHE[key]
+            if expire_at > time.time():
+                return data
+        with Database.get_session() as db:
+            raw_data = db.query(Config.data).where(Config.setting == key).one_or_none()
```
Comment (svanscho) [inline src/gateway/models.py:265]: filter_by

File: src/gateway/models.py [hunk 5]
```diff
@@ -249,52 +249,53 @@ class Config(Base):
...
+            db.add(config_orm)
+            db.commit()
+        Config.CACHE[key] = (value, time.time() + Config.CACHE_EXPIRY_DURATION)
+
+    @staticmethod
+    def remove_entry(key):
+        # type: (str) -> int
+        """ Removes a setting from the DB """
+        with Database.get_session() as db:
+            rows_deleted = db.query(Config).where(Config.setting == key.lower()).delete()
```
Comment (svanscho) [inline src/gateway/models.py:296]: add commit

File: src/gateway/update_controller.py [hunk 6]
```diff
@@ -681,59 +684,59 @@ def _update_energy_firmware(self, firmware_type, target_version, module_address,
...
+                            module.firmware_version = new_version
+                        module.last_online_update = int(time.time())
+                        module.update_success = True
+                    successes += 1
+                except Exception as ex:
+                    individual_logger.exception('Error when updating {0}: {1}'.format(firmware_type, ex))
+                    if module.id is not None:
+                        module.update_success = False
+                    failures += 1
+                db.commit()
```
Comment (svanscho) [inline src/gateway/update_controller.py:739]: same comment

File: src/gateway/update_controller.py [hunk 7]
```diff
@@ -754,7 +757,8 @@ def _filter_modules_to_update(all_modules, target_version, mode):
                     modules_to_update.append(module)
                 else:
                     module.update_success = True
-                    module.save()
+                    db.add(module)
+                    db.save()
```
Comment (svanscho) [inline src/gateway/update_controller.py:761]: db.commit()

File: src/gateway/update_controller.py [hunk 8]
```diff
@@ -744,8 +747,8 @@ def _update_energy_firmware(self, firmware_type, target_version, module_address,
         return successes, failures
 
     @staticmethod
-    def _filter_modules_to_update(all_modules, target_version, mode):
-        # type: (List[Module], str, str) -> List[Module]
+    def _filter_modules_to_update(db, all_modules, target_version, mode):
+        # type: (Session, List[Module], str, str) -> List[Module]
```
Comment (svanscho) [inline src/gateway/update_controller.py:751]: Any

File: testing/unittests/gateway_tests/scheduling_controller_test.py [hunk 9]
```diff
@@ -315,7 +318,7 @@ def _add_schedule(self, **kwargs):
         schedules = self.controller.load_schedules()
         dto = ScheduleDTO(id=None, source='gateway', **kwargs)
         self.controller.save_schedules([dto])
-        for _ in range(60):
+        for _ in range(100):
```
Comment (svanscho) [inline testing/unittests/gateway_tests/scheduling_controller_test.py:321]: revert

**Review by LnL7 — APPROVED** [review]

**Pull request merged**